Enh/npc filtering

cwastche opened the pull request:
- Implements villagers sorting.
- Removed npc types and panels (travelers, residents, atGates)
- Created new Villager type, for a more modular approach
- Refactoring of Villagers Panel and components

edisontim reviewed (commented):
Few comments

cwastche commented on client/src/components/cityview/realm/villagers/panels/villagers/VillagersPanel.tsx:
```diff
@@ -40,79 +37,61 @@ export const VillagersPanel = () => {
+    ),
+  };
+
+  const [selectedNpc, setSelectedNpc] = useState<Npc | undefined>(undefined);
+  const [npcIsSpawning, setNpcIsSpawning] = useState(false);
+
+  const [activeNpcGroup, setActiveNpcGroup] = useState("None");
+  const [activeSort, setActiveSort] = useState<SortInterface>({
```
Yes:
- selectedNpc is used to determine which NPC is being shown in the NpcPopup
- npcIsSpawning is introduced in the previous PR to determine whether an NpcSpawn request has been made, and is used to manage its loading effect
- activeNpcGroup is used to filter the types of npcs (resident, traveler, atGates)
- activeSort is used to filter the npc on a given param (name, age, sex, ...)

edisontim replied:
Can we rename to: 
- `selectedNpc` to `npcDisplayedInPopup`
- `activeNpcGroup` to `displayedNpcGroup`
- `activeSort` to `sortedByParam`

cwastche commented on client/src/components/cityview/realm/villagers/panels/villagers/VillagersPanel.tsx:
```diff
@@ -127,9 +106,59 @@ export const VillagersPanel = () => {
+  const getSpawnInfo = () => {
+    const res = getLastSpawnDelay();
+
+    if (res[0]) {
```
Indeed, i cooked a bit too much on this one.
Changed it to return a single value for better readability.

cwastche commented:
Addressing comments:
- Changed 'res' variable name to 'lastSpawnTs'
- Added a 'READY_TO_SPAWN'
constant for readability
- Changed the return type of getSpawnInfo() to a single value (Either 'READY_TO_SPAWN or spawn delay left)
- Changed name of use... files and functions

edisontim commented on client/src/components/cityview/realm/villagers/panels/villagers/VillagersPanel.tsx:
```diff
@@ -125,16 +125,13 @@ export const VillagersPanel = () => {
-          {/* Format better, mb using interface */}
-          {formatSecondsLeftInDaysHours(Number(res[1]))}
-        </div>
+        <div className="ml-1 text-gold">{formatSecondsLeftInDaysHours(Number(lastSpawnTs))}</div>
```
Conversion to Number necessary? Isn't it already a Number?

edisontim commented on client/src/components/cityview/realm/villagers/panels/villagers/VillagersPanel.tsx:
```diff
@@ -125,16 +125,13 @@ export const VillagersPanel = () => {
   };
 
-  const getLastSpawnDelay = () => {
+  const getLastSpawnTs = () => {
```
Add return type to definition

edisontim commented on client/src/components/cityview/realm/villagers/VillagerComponent.tsx:
```diff
@@ -15,8 +15,8 @@ type VillagerComponentProps = {
   const [showTravel, setShowTravel] = useState(false);
-  const { getNpcInfo } = useVillagerType();
-  const { getExtraButtons } = useVillagerButtons({ villager, setSelectedNpc, setShowTravel });
+  const { getVillagerTypeInfo } = villagerTypeInfo();
+  const { getVillagerButtons } = villagerButtons({ villager, setSelectedNpc, setShowTravel });
```
Remove from a dictionary if there's only one key/value in the dict. Just return the functions directly

edisontim replied:
Also, why return a function in the case of the buttons? Why not directly the value?

edisontim replied:
And why not use directly a function in the case of the info, why do you use a function which returns a function?

edisontim reviewed (approved):
lgtm
